check-meta.nix: Add package warnings (#177272)
NixOS/nixpkgs: doc/using/configuration.chapter.md

-   The package has known security vulnerabilities but has not or can not be updated for some reason, and a list of issues has been entered in to the package's `meta.knownVulnerabilities`.

> -   The package has some other warnings which must be acknowledged, e.g. deprecation notices

Reviewer: ```suggestion
-   The package has some other warnings which must be acknowledged, e.g. deprecation notices.
```

    Note that `permittedInsecurePackages` is only checked if `allowInsecurePredicate` is not specified.

## Installing packages with warnings {#sec-ignore-warnings}

> There are several ways to tweak how Nix handles a package which has some warnings.

Author: I don't think that your proposal is an improvement, but I'm open to other suggestions. Maybe how about "package for which there are warnings"?

> There are several ways to tweak how Nix handles a package which has some warnings.

Reviewer: ```suggestion
There are several ways to tweak how Nix handles a package for which there are warnings.
```

> -   To temporarily allow ignore all warnings, you can use an environment variable for a single invocation of the nix tools:
> 
>     ```ShellSession
>     $ export NIXPKGS_IGNORE_WARNINGS=1
>     ```

Reviewer: `export`ing a variable holds for the shell session. Either make a different example such as

```suggestion
-   To temporarily ignore all warnings, you can use an environment variable for a single invocation of Nix commands:

    ```ShellSession
    $ env NIXPKGS_IGNORE_WARNINGS=1 nix-shell -p hello
    ```
```

or reword

```suggestion
-   To temporarily ignore all warnings, you can use an environment variable for modifying Nix configuration:

    ```ShellSession
    $ export NIXPKGS_IGNORE_WARNINGS=1
    ```
```

Author: I am with you, however I mostly copied the wording over from `pkgs/stdenv/generic/check-meta.nix` messages and changed no more than necessary. So I suggest fixing all of them at once, to remain consistent

Reviewer: In turn, I am with you that you probably don't want to touch unrelated parts, but there is a dilemma: proliferating errors (or confusion, as in this case) in documentation creates more work (and therefore a higher barrier to contributing) for whoever wants to fix this. I know this kind of review comment is super annoying because it tries to pry open your PR's scope. In this case I didn't know, but would still kindly ask you to consider. In my experience, as the manual is so massive, there is usually just one rare chance to iron out kinks like these: drive-by style, while there is some attention to the issue.

> -   It is possible to permanently but selectively ignore all warnings from specific packages, while still blocking other warnings by default using the `ignoreWarningsPackages` configuration option in the user configuration file.

Reviewer: ```suggestion
-   Permanently but selectively ignore all warnings from specific packages, while still showing other warnings by default, with the `ignoreWarningsPackages` configuration option in the [user configuration file](#chap-packageconfig).
```

    The following example configuration permits the installation of the hypothetical package `hello`, version `1.2.3`:

    ```nix
    {
      ignoreWarningsPackages = [
        "hello-1.2.3"
      ];
    }
    ```

> -   It is also possible to create a custom policy around which warnings to ignore, by overriding the `ignoreWarningsPredicate` configuration option.

Reviewer: ```suggestion
-   Create a custom policy as to which warnings to ignore by overriding the `ignoreWarningsPredicate` configuration option.
```

    The following example configuration permits the installation of the hypothetical package `hello`, version `1.2.3`:

    ```nix
    {
      ignoreWarningsPackages = [
        "hello-1.2.3"
      ];
    }
    ```

-   It is also possible to create a custom policy around which warnings to ignore, by overriding the `ignoreWarningsPredicate` configuration option.

    The `ignoreWarningsPredicate` option is a function which accepts a package and returns a boolean, much like `allowUnfreePredicate`.

    The following configuration example ignores all deprecation warnings:

    ```nix
    {
      ignoreWarningsPredicate = pkg: lib.all (issue: issue.kind == "deprecated") (pkg.meta.issues or []);
    }
    ```

    Note that `ignoreWarningsPackages` is only checked if `ignoreWarningsPredicate` is not specified.

### `config` Options Reference

The following attributes can be passed in [`config`](#chap-packageconfig).

```{=docbook}
<include xmlns="http://www.w3.org/2001/XInclude" href="../doc-support/result/config-options.docbook.xml"/>
```

> 

Reviewer 2: This duplicates the existing xinclude, causing the reported build failure.
Might have been caused by a bad rebase.

Author: Oops, my bad. Thanks for noticing

    Note that `permittedInsecurePackages` is only checked if `allowInsecurePredicate` is not specified.

## Installing packages with warnings {#sec-ignore-warnings}

There are several ways to tweak how Nix handles a package for which there are warnings.

-   Temporarily ignore all warnings using an environment variable:

    ```ShellSession
    $ export NIXPKGS_IGNORE_WARNINGS=1
    ```

-   Permanently but selectively ignore some warnings, while still showing others by default, using the `ignoreWarnings` configuration option in the [user configuration file](#chap-packageconfig). Example:

    ```nix
    {
      ignoreWarnings = [

>         "hello.*" # Ignore all warnings emitted from the "hello" package

Reviewer 3: Is this based on name, pname, attrname, ..?

    }
    ```

>     Note that `ignoreWarningsPackages` is only checked if `ignoreWarningsPredicate` is not specified. You can manually integrate the expression `builtins.elem (lib.getName pkg) ignoreWarningsPackages` into the predicate to still make use of the package list in addition to the custom predicates.

Reviewer: Always be explicit and provide full examples. We can make it easier for readers to make use of our instructions.

````suggestion
    Note that `ignoreWarningsPackages` is only used if `ignoreWarningsPredicate` is not specified.
    
    You can manually inspect the list `ignoreWarningsPackages` to combine it with custom predicates, as in the following example:
    
    ```nix
    {
      ignoreWarningsPredicate = pkg: 
        builtins.elem (lib.getName pkg) ignoreWarningsPackages
        ||
        lib.all (issue: issue.kind == "deprecated") (pkg.meta.issues or [])
    }
    ```
````

This is probably a weird way of formatting it. Feel free to come up with something that fits to extending the previous example and is still readable.

Author: I agree it's not ideal. Before applying your suggestion though, I'll try to come up with a solution where users don't have to deal with such thing in the first place.
